File: 快速排序.cpp
// 快速排序

/*
 * 1.确定分界点：q[l], q[(l+r)/2], q[r] 随机   *
 * 2.调整区间，将小于分界点的数放到左边，大于的放到右边
 * 3.递归处理左，右两端
 * */

#include <iostream>
using namespace std;

const int N = 1e5 + 10;
int n;
int a[N];

void quick_sort(int a[], int l, int r)
{
    if (l >= r) return;
    // 判断边界
    int x = a[(l+r)/2], i = l - 1, j = r + 1;
    // x为分界点将i与j设定为初始左右两端的指针，保留原来的l和r用作数据两端的位置
    while (i < j)
    {
        do i++;while (a[i] < x);
        // 查找大于分界点的数
        do j--;while (a[j] > x);
        // 查找小于分界点的数
        if (i < j)
        {
            int temp = a[i];
            a[i] = a[j];
            a[j] = temp;
        }
        // 交换
    }
    // 将剩下的两部分再排序
    quick_sort(a, l, j);
    quick_sort(a, j + 1, r);
    // 此处不是以mid为分界线，mid仅用于随机寻找一个点当分界点，排序完成后以i和j为分界点
}

int main()
{
    scanf("%d", &n);
    for (int i = 0; i < n; i++) scanf("%d", &a[i]);

    quick_sort(a, 0, n-1);
    // 0是最左边的位置，n-1是最右边的位置
    for (int i = 0; i < n; i++) printf("%d ", a[i]);

    return 0;
}

/*
 * 递归实现：
 * 双指针从左右两端向中间靠拢
 * 如果l指针大于分界点，l停止，r指针向中间靠拢，如果r指针小于中间值，l与r对应的值交换
 * */


// 疑问：边界处理？

File: template/快速排序.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from typing import List
from functools import cache, lru_cache
from collections import Counter
from bisect import bisect_right
from itertools import groupby, permutations
from string import ascii_lowercase
from math import inf, sqrt


class Solution:
    arr: List[int] = []

    def quick_sort(self, l: int, r: int) -> None:
        if l >= r:
            return

        mid: int = (l + r) // 2
        x: int = self.arr[mid]

        i, j = l, r
        while i <= j:
            print(i, j)
            while i <= r and self.arr[i] < x:
                i += 1
            while j >= l and self.arr[j] > x:
                j -= 1
            if i <= j:
                self.arr[i], self.arr[j] = self.arr[j], self.arr[i]
                i += 1
                j -= 1

        if l < j:
            self.quick_sort(l, j)
        if i < r:
            self.quick_sort(i, r)

    def sort(self, arr: List[int]) -> List[int]:
        self.arr = arr
        self.quick_sort(0, len(arr) - 1)
        return self.arr


def solve():
    s = Solution()
    print(s.sort([2, 1, 6, 10, 4, 1, 3, 9, 7]))


if __name__ == "__main__":
    import os
    import sys

    if os.getenv("SHELL"):
        infile = open("1.in", "r")
        sys.stdin = infile

    你好 = 1
    # 你好 = int(input())
    for _ in range(你好):
        solve()
